### ISSUE TITLE
Incoming encrypted mail treated as attachment rather than decrypted.

### ISSUE THREAD

**Issue by vasyugan** [nextcloud/mail#3833]
I have installed and configured mailveloped, authorized the nextcloud domain  and enabled the API. On one of my computers yesterday, I saw it working as expected, on the other, I see that incoming PGP encrypted mail (created with Mailvelope) is always displayed only as an attachment, rather than decrypted. 

### Expected behavior
Tell us what should happen
It should just work™

### Actual behavior

![Unbenannt](https://user-images.githubusercontent.com/3313208/96364166-aee54080-1139-11eb-91ca-e2d655c6201b.png)

### Mail app

**Mail app version:** (see apps admin page, e.g. 1.50)

**Mailserver or service:** self-hosted dovecot

### Server configuration

**Operating system**: (e.g. Ubuntu 20.04)

**Web server:** Ningx

**Database:** mariadb

**PHP version:** 7.4

**Nextcloud Version:** 20.0.0

#### Client configuration
**Browser:** Firefox 81

**Operating system:** Debian GNU/Linux Bullseye

**Comment by ChristophWurst** [comment]
The problem is that our client can't handle embedded messages. This is not a problem with encrypted messages but any forwarded or similar message that is not the main part.

Ref https://github.com/nextcloud/mail/pull/2541 for the PR that added the download as attachment.

I *think* there is an existing ticket about the embedded messages but I can't find it right now …

**Comment by vasyugan** [comment]
> The problem is that our client can't handle embedded messages. This is not a problem with encrypted messages but any forwarded or similar message that is not the main part.

Thanks, Christoph, just what irritates me is, with the same browser and the same Mailvelope it works just fine in Roundcube. I wonder if roundcube has some built-in magic that Nextcloud's mail app does not with regard to Mailvelope support. 

Got interested in this because I am in the process of convincing a global human rights network to use Nextcloud and they got unexpectly interested in the mail app, even though they were supposed to  use thunderbird with its build-in PGP support... So they definitely need robust PGP handling. Downloading and manually decoding messages is certainly not an option, as most staff and board are non-technical folks...

**Comment by ChristophWurst** [comment]
> just what irritates me is, with the same browser and the same Mailvelope it works just fine in Roundcube

It's because Mailvelope just detects the PGP message in the browser document and offers decryption. Nextcloud Mail doesn't display the embedded message at all, hence there is nothing for Mailvelope to detect.

So, again, any simple email with an encrypted emails should work fine. It's just the ones that are send embedded as an attachment.

**Comment by vasyugan** [comment]
> > just what irritates me is, with the same browser and the same Mailvelope it works just fine in Roundcube
> 
> It's because Mailvelope just detects the PGP message in the browser document and offers decryption. Nextcloud Mail doesn't display the embedded message at all, hence there is nothing for Mailvelope to detect.
> 
> So, again, any simple email with an encrypted emails should work fine. It's just the ones that are send embedded as an attachment.
No, unfortunately, in my observation. Simple PGP/MIME messages fail, only legacy inline PGP seems to work.

**Comment by ChristophWurst** [comment]
> No, unfortunately, in my observation. Simple PGP/MIME messages fail, only legacy inline PGP seems to work.

Mind exporting such an email as .eml or similar with Thunderbird and posting it here? Then someone can try to reproduce.

**Comment by vasyugan** [comment]
Here are the relevant lines:

```
Content-Type: multipart/encrypted;
 protocol="application/pgp-encrypted";
 boundary="Rfy9LYPuKd9RG8LFoc9aPDXwfzQ32qEQl"

This is an OpenPGP/MIME encrypted message (RFC 4880 and 3156)
--Rfy9LYPuKd9RG8LFoc9aPDXwfzQ32qEQl
Content-Type: application/pgp-encrypted
Content-Description: PGP/MIME version identification

Version: 1

--Rfy9LYPuKd9RG8LFoc9aPDXwfzQ32qEQl
Content-Type: application/octet-stream; name="encrypted.asc"
Content-Description: OpenPGP encrypted message
Content-Disposition: inline; filename="encrypted.asc"

-----BEGIN PGP MESSAGE-----

wcBMA0AiiRyUAYVDAQf+OUZ3A2Yrpw+EhfFIr00lpkyP6bgrZwoHiolhZvD3zocq9tMRSdDngJg0
gHrgJc91y3jaLSeDuAQUWPU+k4ShevVDM+Z5S2SFwPhRWb5xXj/IANtWPmF4iLjsYSH0tMBpJm/w
```

**Comment by ChristophWurst** [comment]
But that is a multipart message with a dedicated php part. So I guess it's the very same described limitation that Mail can't handle those. If there is a plain text part (as per content type) that contains pgp then it will display, I think.

As a remedy we could always show the `application/pgp-encrypted` part as if it were text and hope that Mailvelope picks it up.

**Comment by vasyugan** [comment]
> But that is a multipart message with a dedicated php part. So I guess it's the very same described limitation that Mail can't handle those. If there is a plain text part (as per content type) that contains pgp then it will display, I think.
> 
> As a remedy we could always show the `application/pgp-encrypted` part as if it were text and hope that Mailvelope picks it up.

As far as I see it, that's the [PGP/Mime standard](https://de.wikipedia.org/wiki/PGP/MIME), which is the default   e.g. GpgOL and Thunderbird. Inline PGP seems to be mainly used by Mailvelope and as I understand, it is legacy. PGP/MIME is the correct™ way, that any mailer should support, at least passively.
 

Just to be sure, this here is from a message composed in Thunderbird 78:
```

This is an OpenPGP/MIME encrypted message (RFC 4880 and 3156)
--siAL4TRdtcJVH51ByjyPKQ6ZH8U234sU2
Content-Type: application/pgp-encrypted
Content-Description: PGP/MIME version identification

Version: 1

--siAL4TRdtcJVH51ByjyPKQ6ZH8U234sU2
Content-Type: application/octet-stream; name="encrypted.asc"
Content-Description: OpenPGP encrypted message
Content-Disposition: inline; filename="encrypted.asc"

-----BEGIN PGP MESSAGE-----

wcBMA0AiiRyUAYVDAQf9F77iZuyWvXnXWHJzX1TgqLxj+eu8OZvwkRgJLCIIq/xS98raHnY/Il+P
Sq2pMc6NEKjvAOIPEm/jJPhI09Jt9dtoqMRLEqlAts29XP8YU7vWgCR2So4aPVu9/hRNm6lwfD2s 
```

**Comment by ChristophWurst** [comment]
Thanks for the info! I will look into this when time allows but right now this is a very low priority. If you are a developer or know someone who could help with the development I'm happy to discuss what we might need to change.

**Comment by vasyugan** [comment]
> Thanks for the info! I will look into this when time allows but right now this is a very low priority. If you are a developer or know someone who could help with the development I'm happy to discuss what we might need to change.

Unfortunately, I am  not a developer, my modest contributions to Nextcloud are limited to documentation. However, I should flag that as long as the combo of Nextcloud Mail and Mailvelope cannot handle PGP/MIME messages, it does not implement the established standard, and in my case, this means that it cannot be used by the network as their default mail client, which is a shame because else, they will be using nextcloud for most of their needs.

There is a long-standing Mailvelope bug related to PGP/MIME, https://github.com/mailvelope/mailvelope/issues/41 but I guess that fixing this issue here would not have to wait this other issue be fixed, because all Nextcloud Mail would have to do is download the bodies of messages with the content type "application/pgp-encrypted".

I am interacting with some other organisations who are focused on cybersecurity  for civil society and might therefore be interested in contributing, but I would not be optimistic, that they have any free resources to devote to this, unfortunately.

**Comment by ChristophWurst** [comment]
No worries @vasyugan, I appreciate the time you take to report this.

I agree that it's not too much effort. But this is one of many small changes that we have to work on. And everyone has this ticket where they think it's *that* most important one. But of course that always depends on the user context.

In any case, leaving this open for pickup later or anyone else who would like to help.

Any enterprises using the software can get a subscription and we'll be then able to shift priorities accordingly. But I guess that doesn't work for you organizations.

**Comment by vasyugan** [comment]
> No worries @vasyugan, I appreciate the time you take to report this.
> 
> I agree that it's not too much effort. But this is one of many small changes that we have to work on. And everyone has this ticket where they think it's _that_ most important one. But of course that always depends on the user context.
> 
> In any case, leaving this open for pickup later or anyone else who would like to help.
> 
> Any enterprises using the software can get a subscription and we'll be then able to shift priorities accordingly. But I guess that doesn't work for your organizations.

Indeed. Civil society networks of that sort have limited resources anyway, and the situation was that I was trying to evangelize them to use PGP and then they discovered that Nextcloud comes with its own mail client which they would prefer to use over Thunderbird, and now I am looking for ways to reconcile security needs with their longing for convenience...

**Comment by vasyugan** [comment]
> But that is a multipart message with a dedicated php part. So I guess it's the very same described limitation that Mail can't handle those. If there is a plain text part (as per content type) that contains pgp then it will display, I think.
> 
> As a remedy we could always show the `application/pgp-encrypted` part as if it were text and hope that Mailvelope picks it up.

I think that's all that Roundcube does and that indeed suffices to tackle the situation.

**Comment by vasyugan** [comment]
@ChristophWurst I just grepped through the roundcube source code for encrypted.asc and application/pgp-encrypted, and it turns out, that it doesn't even have any special code for downloaded and displaying pgp/mime parts. It just seems to honour "Content-Disposition: Inline". Can't Nextcloud Mail just do the same?

**Comment by ChristophWurst** [comment]
> It just seems to honour "Content-Disposition: Inline". Can't Nextcloud Mail just do the same?

"just". Yeah …

**Comment by vasyugan** [comment]
> > It just seems to honour "Content-Disposition: Inline". Can't Nextcloud Mail just do the same?
> 
> "just". Yeah …

I see that lib/Attachment.php explicitly changes the disposition of mime parts from inline to attachment
The comment says: _"Serve all files with a content-disposition of "attachment" to prevent Cross-Site Scripting"_

So I see, this is a precautionary measure, but it is one that  roundcube apparently does not need.

(I thought I coud just make it work by commenting out this line, but this doesn't seem to have been enough)

 given that it prevents PGP from functioning properly, it seems to be worth it, exploring alternatives. Again, sadly, I am not a developer, but the task seems relatively limited: ensure that XSS is prevented while content-disposition: inline parts are indeed displayed inline.

Would this be a correct description of the issue?

**Comment by ChristophWurst** [comment]
It's more complex than that. It's not just what part to hide or what not to show. It can become this recursive structure where you have a mesage inside a message inside a message.

This has to be done properly.

So it's blocked until we found the time to work on this or a customer funds the development.

**Comment by gpgmailencrypt** [comment]
May be a bit late but I just want to clarify something about the PGP/MIME format. 
A PGP/MIME message just has one attachment, the enrypted message. Once decrypted it contains the message body and all attachments, just like any other non-encrypted message. It even can contain some of the header data like the subject.

The advantage of PGP/Mime over pgp/inline is that it hides the complete content (including file names of attachments) and it keeps all formatting of the message body. See https://tools.ietf.org/html/rfc3156 for details. 

And if you like you might want to have a look in the sourcecode of my software https://github.com/gpgmailencrypt/gpgmailencrypt/blob/master/gpgmailencrypt.py#L5001
You simply need to combine the headers of the encrypted email ( escpecially FROM, RECEIVED etc) with the decrypted body. Just the subject needs some additional attention, as it may be part of the encrypted message body (see above) 

Technically a mail could contain a pgp/mime attachment and different non-encrypted or pgp-inline encrypted attachments. But this would be totally against the idea of PGP/MIME and I know no e-mail client that can handle such an e-mail correctly.

**Comment by ioogithub** [comment]
> So it's blocked until we found the time to work on this or a customer funds the development.
I am not an enterprise customer however I may be interested in funding this development. Who do I speak to about this? It is important because without it, Nextcloud and mailvelope mail does not work at all for me.

**Comment by vasyugan** [comment]
> I am not an enterprise customer however I may be interested in funding this development. Who do I speak to about this? It is important because without it, Nextcloud and mailvelope mail does not work at all for me.

@ioogithub i think one would have to open a bounty for third party developers as nextcloud only works with customers through enterprise subscriptions

**Comment by ChristophWurst** [comment]
https://help.nextcloud.com/c/nextcloud-freelancing/48 is the place to talk about freelance funding. Bounties are a bit of a dated concept right now: https://help.nextcloud.com/t/dont-use-bountysource-anymore/84943

**Comment by vasyugan** [comment]
> https://help.nextcloud.com/c/nextcloud-freelancing/48 is the place to talk about freelance funding. Bounties are a bit of a dated concept right now: https://help.nextcloud.com/t/dont-use-bountysource-anymore/84943

Thanks! For me, I would have to see where to get external funding for this. There are donors funding cybersecurity related applications, but I would have to go looking. But before that, I guess one should have an idea how much money is actually needed, and also, what the freelancer needs to do to ensure that the patches are accepted by Nextcloud, @ChristophWurst is there any quick read on this? @ioogithub Did I correctly understand you that you might be willing to fund this from your private pocket?

**Comment by ChristophWurst** [comment]
In general I strongly agree with this plan @vasyugan. This week and next I won't have time to look into the details in order to provide any action items or a list of things to change.

**Comment by vasyugan** [comment]
> In general I strongly agree with this plan @vasyugan. This week and next I won't have time to look into the details in order to provide any action items or a list of things to change.

Do you imply that you will eventually get back to us with a list of changes that would have to be made? That certainly would make the process much easier, as we could specifically tell the freelancer what he needs to change.

**Comment by ChristophWurst** [comment]
I'll try but I can't promise :v:

**Comment by vasyugan** [comment]
So I'll nag you again two weeks from now, I guess.

Am 22.03.21 um 12:55 schrieb Christoph Wurst:

**Comment by ioogithub** [comment]
>  @ioogithub Did I correctly understand you that you might be willing to fund this from your private pocket?

Yes I am interested in funding this because without it I can't use Nextcloud Mail and so without it, a cloud platform that can't do email is much less useful.

From the discussion above it sounds like bountysource is no longer a viable option. I contacted Nextcloud sales and although they thanked me for the offer of support they confirmed that they do not accept commissioned feature development.

They also mentioned specifically that this request is not a priority so there is a real risk of funding freelance development work which will not be accepted by the company and integrated into the product for various reasons. Unless I am missing something, it doesn't sound like there is any acceptable mechanisms in place for integrating new features with this funding model.

**Comment by vasyugan** [comment]
> > @ioogithub Did I correctly understand you that you might be willing to fund this from your private pocket?
> 
> Yes I am interested in funding this because without it I can't use Nextcloud Mail and so without it, a cloud platform that can't do email is much less useful.
> 
> From the discussion above it sounds like bountysource is no longer a viable option. I contacted Nextcloud sales and although they thanked me for the offer of support they confirmed that they do not accept commissioned feature development.
> 
> They also mentioned specifically that this request is not a priority so there is a real risk of funding freelance development work which will not be accepted by the company and integrated into the product for various reasons. Unless I am missing something, it doesn't sound like there is any acceptable mechanisms in place for integrating new features with this funding model.

Maybe a viable option would be that a freelancer rather than patching the mail app itself develops the needed additions in form of another app that extends the functionality of the mail app.

Anyway, I think we need to keep nagging @ChristophWurst for directions.

Yes, it is a tragedy that e-mail encryption gets so little love, even from Nextcloud. Unencrypted email in my view is the most glaring security hole. Yes, I agree that for instance Telegram's security model leaves much to be desired, but normal unencrypted e-mail has no security model at all. It is basically as it was 50 years ago.

**Comment by vasyugan** [comment]
> I'll try but I can't promise 

@ChristophWurst Maybe, do you have time now to give us advice?

**Comment by vasyugan** [comment]
> 
> 
> I'll try but I can't promise ✌️

Hey, I thought it was time to poke you again. At present, PGP support is broken, as the one standard, which is state of the art is not supported. If Nextcloud GmbH does not see this as a priority, it would be great if you could give us directions on what we have to to go 
1) hire  a third party to fix it 
2) ensure that their additions are accepted by Nextcloud
Does it make sense to go for a separate app that extends the mail app's functionality?

**Comment by ioogithub** [comment]
>     1. hire  a third party to fix it
I offered to fund it personally. Nextcloud sales said they would not accept this.

As recommended, I posting a freelancing ad, it got 0 replies.

>     2. ensure that their additions are accepted by Nextcloud
This seems to be the real problem with Nextcloud. I just saw an issue where a user found a verifiable bug, did a deep dive, coded and tested the solution. The fix was rejected because it was judged to be a bug that would not effect enough users... The reason given was that they do not have the resources to integrate the fix.

So Nextcloud rejects code solutions and also funding which they could use to integrate code solutions and improve the project.

I have to admit I do not understand how this model will lead to success.

**Comment by ChristophWurst** [comment]
@vasyugan I haven't forgotten. I'll get back to you soon.

**Comment by gpgmailencrypt** [comment]
To make it easier for a programmer here is a visual example how pgpmime works. If you need more information let me know
[pgpmime.pdf](https://github.com/nextcloud/mail/files/6411058/pgpmime.pdf)

**Comment by vasyugan** [comment]
> @vasyugan I haven't forgotten. I'll get back to you soon.

Really sorry for all the nagging, but here is another friendly reminder...

**Comment by vasyugan** [comment]
> 
> 
> @vasyugan I haven't forgotten. I'll get back to you soon.

Here is another reminder. Still friendly, but already a bit impatient...

**Comment by ChristophWurst** [comment]
> To make it easier for a programmer here is a visual example how pgpmime works. If you need more information let me know
> [pgpmime.pdf](https://github.com/nextcloud/mail/files/6411058/pgpmime.pdf)

Thanks a lot @gpgmailencrypt. The detail with `Content-Type: multipart/encrypted; protocol="application/pgp-encrypted";` is very helpful, if that is set consistently by all clients then the detection of gpg/mime should be easier.

On this second thought it also looks like we don't even have to fix the limiter of embedded message rending in Mail for this specific feature. If That content-type and the protocol is detected, we can discard the first part and only show the `application/octet-stream` but as plain text. This should be possible to pick up by Mailvelope then.

Now to the more technical part. Right now most of the logic is located inside `\OCA\Mail\Model\IMAPMessage::loadMessageBodies`. Touching this can work but is dangerous, it's one of the remaining legacy parts of this app. New IMAP code is generally moved over to `\OCA\Mail\IMAP\MessageMapper` and `loadMessageBodies` should likely be migrated as well so the feature can get reliable unit and integration tests.

Right now the boundary part of the encrypted message is ignored, which is good. The main part of the message is however detected as attachment because ``\Horde_Mime_Part::isAttachment`` return `true`. There is handling for `multipart/encrypted` in that method but `$this->parent` is `null`.

With that information I'd say the actionable items are

* Refactor ``\OCA\Mail\Model\IMAPMessage::getPart`` into a new method of ``\OCA\Mail\IMAP\MessageMapper`` and unit test all the existing logic
* Find out why Horde's logic for attachments fails on those messages
* Extend the new method in the mapper so that it can cope with ``multipart/encrypted`` if Horde doesn't do it already when used correctly
* Send the encrypted part as plain text message to the front-end as if it were text
* Show a warning to the user if Mailvelope is not there but such a part is displayed, otherwise leave the rest to Mailvelope

**Comment by vasyugan** [comment]
> > To make it easier for a programmer here is a visual example how pgpmime works. If you need more information let me know
> > [pgpmime.pdf](https://github.com/nextcloud/mail/files/6411058/pgpmime.pdf)
> 
> Thanks a lot @gpgmailencrypt. The detail with `Content-Type: multipart/encrypted; protocol="application/pgp-encrypted";` is very helpful, if that is set consistently by all clients then the detection of gpg/mime should be easier.

Since this is what the standard prescribes, this should be a given. Just checked messages generated by Thunderbird, Outlook (with the GPG OL extension) and by Protonmail. They all conform to the standard
[...]

> With that information I'd say the actionable items are
> 
>     * Refactor `\OCA\Mail\Model\IMAPMessage::getPart` into a new method of `\OCA\Mail\IMAP\MessageMapper` and unit test all the existing logic
> 
>     * Find out why Horde's logic for attachments fails on those messages
> 
>     * Extend the new method in the mapper so that it can cope with `multipart/encrypted` if Horde doesn't do it already when used correctly
> 
>     * Send the encrypted part as plain text message to the front-end as if it were text
> 
>     * Show a warning to the user if Mailvelope is not there but such a part is displayed, otherwise leave the rest to Mailvelope

Thanks a lot! For me as a layman, this all sounds like something that cannot realistically be implemented as a separate app, but something that requires patches to the mail app, which means that whoever takes on the task will have to interact with the core developers and if and when the patches are accepted will ultimately be outside of their control, it may hinge on factors such as the development roadmap, right? Would it make sense to take a two-pronged approach such as by forking the mail app and at the same time trying to reintegrate the changes into the upstream code?

**Comment by vasyugan** [comment]
> > ```
> > 1. hire  a third party to fix it
> > ```
> 
> I offered to fund it personally. Nextcloud sales said they would not accept this.
> 
> As recommended, I posting a freelancing ad, it got 0 replies.
Where did you post it?

**Comment by ChristophWurst** [comment]
> requires patches to the mail app, which means that whoever takes on the task will have to interact with the core developers and if and when the patches are accepted will ultimately be outside of their control, it may hinge on factors such as the development roadmap, right? Would it make sense to take a two-pronged approach such as by forking the mail app and at the same time trying to reintegrate the changes into the upstream code?

I do not have any objections with those changes in this app as long as all existing multi part handling works as before. So I'll happily accept any contribution that follows this rough plan that I proposed. What I would reject is patches that just *make it work* but without any kind of testing that helps us ensure the stability of existing features.

**Comment by vasyugan** [comment]
Thanks @ChristophWurst Unfortunately, since our partners are now saying that they are using Thunderbird, not only because of its working PGP/MIME support, we no longer have a grounds for raising the funds needed to pay for the development work. So I guess, unless someone else takes it up, this bug will remain unresolved.

**Comment by NoSubstitute** [comment]
@ChristophWurst - has this issue just been forgotten?
Asking because I just used Mailvelope to successfully open and read an encrypted email sent from Mailvelope.

Sent it from my Gmail account to NC with Mailvelope installed in Chrome. Both webmail services opened in separate tabs in the same Chrome profile.

**Comment by vasyugan** [comment]
> @ChristophWurst - has this issue just been forgotten? Asking because I just used Mailvelope to successfully open and read an encrypted email sent from Mailvelope.
> 
> Sent it from my Gmail account to NC with Mailvelope installed in Chrome. Both webmail services opened in separate tabs in the same Chrome profile.

I think someone has to pick it up @ChristophWurst  has described the tasks in [his comment](https://github.com/nextcloud/mail/issues/3833#issuecomment-855709468).

**Comment by NoSubstitute** [comment]
For sure, but the problem description doesn't seem to be true anymore, or I am doing something wrong/right?
I just used Mailvelope with NC Mail and it worked fine, no issues.

I created an encrypted email with Mailvelope (on Gmail) and sent it to my NC account.
I opened the encrypted email with Mailvelope in NC mail, without issues.

**Comment by vasyugan** [comment]
> For sure, but the problem description doesn't seem to be true anymore, or I am doing something wrong/right? I just used Mailvelope with NC Mail and it worked fine, no issues.
> 
> I created an encrypted email with Mailvelope (on Gmail) and sent it to my NC account. I opened the encrypted email with Mailvelope in NC mail, without issues.

That's because Mailvelope uses legacy inline PGP, rather than PGP/MIME. Try to open a message sent by e.g. Thunderbird or MS Outlook with the GPGOL add-on, and you will still see it fail. All full-fledged mail clients such as Thunderbird or Outlook with GpgOL use PGP/MIME. See [this page](https://proton.me/support/pgp-mime-pgp-inline) for an explanation of the differences.

**Comment by NoSubstitute** [comment]
Oh, it seems to be not only that.
Recent version of Engimail with Interlink Email seems to **not include who the recipient is**. Meaning the info about which key is supposed to be used when decryption _isn't included in the encrypted content_, so the decrypting software has to guess!
<img width="414" alt="Enigmail_fails_too" src="https://user-images.githubusercontent.com/12572734/179215201-4729756a-b253-4a30-b2fd-2d66039b3e0b.png">

With old Enigmail on old Thunderbird (still compatible with Enigmail) that means it has to ask for the password to all private keys in the attempt to decrypt the content, and until it works, it just asks for the password to the next key.

I thought that Old Enigmail on old Thunderbird would include info on which the recipient key is, and therefore be opened by Mailvelope, but that too doesn't include that information.

Only Mailvelope includes the recipient key info.
<img width="319" alt="Only_Mailvelope_includes_recipient_key_info" src="https://user-images.githubusercontent.com/12572734/179215144-cc3f2115-b057-4894-b004-fd2e94726261.png">

Mailvelope doesn't have the guessing feature, and instead just says _no private key included_ and then does nothing.
<img width="385" alt="No_private_key_included" src="https://user-images.githubusercontent.com/12572734/179215094-dec42e5a-983b-421a-8cc3-58a727ea22f7.png">

This change of not including the recipient key breaks Mailvelope, regardless of email service (afaik, at least Gmail anc Nextcloud), and also "breaks" email client software like old Thunderbird and new Interlink (both using their respective version of Enigmail).

**lol, and just now I found out why this is happening!**
It's a setting in gpg!
[throw-keyid](https://www.gnupg.org/gph/en/manual/r2110.html) - seems it can be used both with and without an s at the end (throw-keyids).

Disabling throw-keyid makes PGP-INLINE work great from other clients, readable by Mailvelop on NC Mail.

But I have now also verified that PGP/MIME doesn't work with NC Mail & Mailvelope, because NC Mail doesn't show a preview of the attachment. Just like @ChristophWurst said early on.

NC Mail doesn't display attachments/embedded content, and then Mailvelope has nothing to read.

So, it's bad, but it's not useless.

**Comment by NoSubstitute** [comment]
I meant that afaik all the '"real" mail clients' have the option of sending inline, and don't have to use PGP/MIME.

It's not that big of a hassle to tell those _extremely_ few people who send PGP (/MIME) messages to please use inline when sending to you. There's probably even an address book setting where it can be set per recipient. At least if it is a "real" client. ;-)

**Comment by mdengler** [comment]
👍 🍾
